libraries/AP_Mount/AP_Mount_Viewpro.cpp

```diff
@@ -0,0 +1,782 @@
+    a1_packet.content.servo_status = ServoStatus::manual_speed_mode;
+    a1_packet.content.yaw_be = htobe16(yaw_rate_output);
+    a1_packet.content.pitch_be = htobe16(pitch_rate_output);
+
+    // send targets to gimbal
+    return send_packet(a1_packet.bytes, sizeof(a1_packet.bytes));
+}
+
+// send target pitch and yaw angles to gimbal
+// yaw_is_ef should be true if yaw_rad target is an earth frame angle, false if body_frame
+bool AP_Mount_Viewpro::send_target_angles(float pitch_rad, float yaw_rad, bool yaw_is_ef)
+{
+    // gimbal does not support lock in angle control mode
+    if (!set_lock(false)) {
+        return false;
+    }
+
+    // convert yaw angle to body-frame
+    float yaw_bf_rad = yaw_is_ef ? wrap_PI(yaw_rad - AP::ahrs().yaw) : yaw_rad;
+
+    // enforce body-frame yaw angle limits.  If beyond limits always use body-frame control
+    const float yaw_bf_min = radians(_params.yaw_angle_min);
+    const float yaw_bf_max = radians(_params.yaw_angle_max);
+    if (yaw_bf_rad < yaw_bf_min || yaw_bf_rad > yaw_bf_max) {
+        yaw_bf_rad = constrain_float(yaw_bf_rad, yaw_bf_min, yaw_bf_max);
+        yaw_is_ef = false;
+    }
+
+    // scale pitch and yaw to values gimbal understands
+    const int16_t pitch_angle_output = -degrees(pitch_rad) * AP_MOUNT_VIEWPRO_DEG_TO_OUTPUT;
+    const int16_t yaw_angle_output = degrees(yaw_bf_rad) * AP_MOUNT_VIEWPRO_DEG_TO_OUTPUT;
+
+    // fill in packet
+    A1Packet a1_packet {};
+    a1_packet.content.frame_id = FrameId::A1;
+    a1_packet.content.servo_status = ServoStatus::manual_absolute_angle_mode;
+    a1_packet.content.yaw_be = htobe16(yaw_angle_output);
+    a1_packet.content.pitch_be = htobe16(pitch_angle_output);
+
+    // send targets to gimbal
+    return send_packet(a1_packet.bytes, sizeof(a1_packet.bytes));
+}
+
+// send camera command and corresponding value (e.g. zoom speed)
+bool AP_Mount_Viewpro::send_camera_command(CameraCommand cmd, uint8_t value)
+{
+    // fill in packet
+    C1Packet c1_packet {};
+    c1_packet.content.frame_id = FrameId::C1;
+
+    // fill in 2 bytes containing sensor, zoom speed, operation command and LRF
+    // bit0~2: sensor
+    // bit3~5: zoom speed
+    // bit6~12: operation command no
+    // bit13~15: LRF command (unused)
+    const uint16_t sensor_id = (uint16_t)ImageSensor::EO1;
+    const uint16_t zoom_speed = ((uint16_t)value & 0x07) << 3;
+    const uint16_t operation_cmd = ((uint16_t)cmd & 0x7F) << 6;
+    c1_packet.content.sensor_zoom_cmd_be =  htobe16(sensor_id | zoom_speed | operation_cmd);
+
+    // send packet to gimbal
+    return send_packet(c1_packet.bytes, sizeof(c1_packet.bytes));
+}
+
+// send camera command2 and corresponding value (e.g. zoom as absolute value)
+bool AP_Mount_Viewpro::send_camera_command2(CameraCommand2 cmd, uint16_t value)
+{
+    // fill in packet
+    C2Packet c2_packet {};
+    c2_packet.content.frame_id = FrameId::C2;
+    c2_packet.content.cmd = cmd;
+    c2_packet.content.value_be = htobe16(value);
```

Comment:
Any reason to now assemble the packet entirely rather than doing assignment? 

e.g.
```suggestion
    const C2Packet c2_packet {
         frame_id: FrameId::C2;
         cmd: cmd, 
         value_be: value,
    };
```

Reply:
@peterbarker, thanks for the suggestion. I don't think this is possible because it is a union..